pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [error with DROP VALUE followed by DROP SCHEMA in same txn]

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14707490?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14707490?buildTab=artifacts#/) on release-23.2 @ [914e28276c93fe4b6d56caa6873ff4d63fc4599f](https://github.com/cockroachdb/cockroach/commits/914e28276c93fe4b6d56caa6873ff4d63fc4599f):


```
1, 29, protected_ts_meta, 31
1, 29, protected_ts_records, 32
1, 29, rangelog, 13
1, 29, region_liveness, 9
1, 29, replication_constraint_stats, 25
1, 29, replication_critical_localities, 26
1, 29, replication_stats, 27
1, 29, reports_meta, 28
1, 29, role_id_seq, 48
1, 29, role_members, 23
1, 29, role_options, 33
1, 29, scheduled_jobs, 37
1, 29, settings, 6
1, 29, span_configurations, 47
1, 29, span_stats_buckets, 55
1, 29, span_stats_samples, 56
1, 29, span_stats_tenant_boundaries, 57
1, 29, span_stats_unique_keys, 54
1, 29, sql_instances, 46
1, 29, sqlliveness, 39
1, 29, statement_activity, 60
1, 29, statement_bundle_chunks, 34
1, 29, statement_diagnostics, 36
1, 29, statement_diagnostics_requests, 35
1, 29, statement_execution_insights, 65
1, 29, statement_statistics, 42
1, 29, table_statistics, 20
1, 29, task_payloads, 58
1, 29, tenant_id_seq, 62
1, 29, tenant_settings, 50
1, 29, tenant_tasks, 59
1, 29, tenant_usage, 45
1, 29, tenants, 8
1, 29, transaction_activity, 61
1, 29, transaction_execution_insights, 64
1, 29, transaction_statistics, 43
1, 29, ui, 14
1, 29, users, 4
1, 29, web_sessions, 19
1, 29, zones, 5
100, 0, public, 101
102, 0, public, 103
104, 0, public, 105
104, 105, _enum54, 117
104, 105, enum54, 116
104, 105, table39, 115
    schema_change_external_test.go:82: backup in /artifacts/tmp/_tmp/fcba020ca7cb44bf8385bd0e13bcbe2d/logTestWorkload3831974478
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/fcba020ca7cb44bf8385bd0e13bcbe2d/logTestWorkload3831974478
--- FAIL: TestWorkload (6.78s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-37574

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14721841?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/14721841?buildTab=artifacts#/) on release-23.2 @ [914e28276c93fe4b6d56caa6873ff4d63fc4599f](https://github.com/cockroachdb/cockroach/commits/914e28276c93fe4b6d56caa6873ff4d63fc4599f):


```
1, 29, namespace, 30
1, 29, privileges, 51
1, 29, protected_ts_meta, 31
1, 29, protected_ts_records, 32
1, 29, rangelog, 13
1, 29, region_liveness, 9
1, 29, replication_constraint_stats, 25
1, 29, replication_critical_localities, 26
1, 29, replication_stats, 27
1, 29, reports_meta, 28
1, 29, role_id_seq, 48
1, 29, role_members, 23
1, 29, role_options, 33
1, 29, scheduled_jobs, 37
1, 29, settings, 6
1, 29, span_configurations, 47
1, 29, span_stats_buckets, 55
1, 29, span_stats_samples, 56
1, 29, span_stats_tenant_boundaries, 57
1, 29, span_stats_unique_keys, 54
1, 29, sql_instances, 46
1, 29, sqlliveness, 39
1, 29, statement_activity, 60
1, 29, statement_bundle_chunks, 34
1, 29, statement_diagnostics, 36
1, 29, statement_diagnostics_requests, 35
1, 29, statement_execution_insights, 65
1, 29, statement_statistics, 42
1, 29, table_statistics, 20
1, 29, task_payloads, 58
1, 29, tenant_id_seq, 62
1, 29, tenant_settings, 50
1, 29, tenant_tasks, 59
1, 29, tenant_usage, 45
1, 29, tenants, 8
1, 29, transaction_activity, 61
1, 29, transaction_execution_insights, 64
1, 29, transaction_statistics, 43
1, 29, ui, 14
1, 29, users, 4
1, 29, web_sessions, 19
1, 29, zones, 5
100, 0, public, 101
102, 0, public, 103
104, 0, public, 105
104, 105, seq123, 136
    schema_change_external_test.go:82: backup in /artifacts/tmp/_tmp/fcba020ca7cb44bf8385bd0e13bcbe2d/logTestWorkload2223719728
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/fcba020ca7cb44bf8385bd0e13bcbe2d/logTestWorkload2223719728
--- FAIL: TestWorkload (12.02s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/drop_type.go
+++ b/pkg/sql/drop_type.go
@@ -292,6 +292,17 @@ func (p *planner) dropTypeImpl(
 		return err
 	}
 
+	// For this type descriptor, delete queued schema change jobs from the job
+	// cache. If the job had already been scheduled, we would also need to mark
+	// these jobs as successful, but that should never happen for types.
+	//
+	// Note that we still wait for jobs removed from the cache to finish running
+	// after the transaction, since they're not removed from the jobsCollection.
+	// Also, changes made here do not affect schema change jobs created in this
+	// transaction with no mutation ID; they remain in the cache, and will be
+	// updated when writing the job record to drop the table.
+	delete(p.ExtendedEvalContext().jobs.uniqueToCreate, typeDesc.ID)
+
 	if err := p.removeBackRefsFromAllTypesInType(ctx, typeDesc); err != nil {
 		return err
 	}

--- a/pkg/sql/jobs_collection.go
+++ b/pkg/sql/jobs_collection.go
@@ -38,7 +38,7 @@ type txnJobsCollection struct {
 	// job for a relation in one transaction. These jobs will be created and
 	// queued at commit time.
 	uniqueToCreate map[descpb.ID]*jobs.Record
-	// uniqueToCreate contains job records that are not unique to a descriptor
+	// nonUniqueToCreate contains job records that are not unique to a descriptor
 	// IDs. These jobs will be created and queued at commit time.
 	nonUniqueToCreate []*jobs.Record
 }

--- a/pkg/sql/logictest/testdata/logic_test/drop_schema
+++ b/pkg/sql/logictest/testdata/logic_test/drop_schema
@@ -6,3 +6,18 @@ CREATE DATABASE test2
 
 statement error pq: cannot drop schema "public"
 DROP SCHEMA test2.public
+
+# Regression test for ALTER TYPE ... DROP VALUE followed by DROP SCHEMA CASCADE.
+# The type schema change should never be executed, since the DROP SCHEMA would
+# have already removed the type.
+statement ok
+CREATE SCHEMA schema_123539;
+
+statement ok
+CREATE TYPE schema_123539.enum_123539 AS ENUM ('s', 't');
+
+statement ok
+BEGIN;
+ALTER TYPE schema_123539.enum_123539 DROP VALUE 's';
+DROP SCHEMA schema_123539 CASCADE;
+COMMIT;

